File: house-robber/jaejeong1.java
class Solution {
  public int rob(int[] nums) {
    // 인접한 경우는 제외한 최대 합
    // 풀이: dp로 풀이한다. dp[i] = max(dp[i-1], dp[i-2] + nums[i])
    // TC: O(N), SC: O(N)
    if (nums.length == 1) {
      return nums[0];
    }

    var dp = new int[nums.length];
    dp[0] = nums[0];
    dp[1] = Math.max(nums[0], nums[1]);

    for (int i=2; i<nums.length; i++) {
      dp[i] = Math.max(dp[i-1], dp[i-2] + nums[i]);
    }

    return dp[nums.length-1];
  }
}

File: longest-repeating-character-replacement/jaejeong1.kt
import kotlin.math.max

fun characterReplacement(s: String, k: Int): Int {
	// 시간복잡도: O(N) 공간복잡도: O(1)

	val frequencyMap = mutableMapOf<Char, Int>()
	var left = 0
	var maxFreq = 0
	var result = 0
	// 처음부터 끝까지 right 순회
	for (right in s.indices) {
		// right 위치 문자에 빈도를 증가시키고 map 저장
		val rightChar = s[right]
		frequencyMap[rightChar] = frequencyMap.getOrDefault(rightChar, 0) + 1
		maxFreq = max(maxFreq, frequencyMap[rightChar]!!)

		val windowSize = right - left + 1
		val downSizeNeeded = windowSize - maxFreq

		if (downSizeNeeded > k) {
			val leftChar = s[left]
			frequencyMap[leftChar] = frequencyMap[leftChar]!! - 1
			left++
		}
		result = max(result, right - left + 1)
	}
}